src/librustc/ty/layout.rs

```diff
@@ -951,19 +1024,20 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                                 Some(niche) => niche,
                                 _ => continue,
                             };
+                            // Do not use a niche smaller than the tag unless it reduces the size.
+                            if !must_grow && niche.available(dl) >> ity.size().bits() == 0 {
```

Comment:
The size of the discriminant shouldn't be used (since it ignores the actual available range). Instead, you want to compare with `.available(dl)` on the tag niche (computed further down):
```rust
let largest_niche = Niche::from_scalar(dl, Size::ZERO, tag.clone());
```